prov/sm2/src/sm2.h

```diff
@@ -308,6 +319,16 @@ static inline size_t sm2_pop_xfer_entry(struct sm2_ep *ep,
 	return FI_SUCCESS;
 }
 
+static inline bool sm2_proto_imm_send_comp(uint16_t proto)
```

Comment:
I haven't seen where this is used yet but it seems like a rather silly function and implementation for
`proto == sm2_proto_cma`
This also seems confusing given the earlier implementation of the generate completion flag/delivery complete flag. not all non cma will be able to generate an immediate send completion, right?
maybe it becomes more important in other patches or you're anticipating other protocols but for now I would drop this function and add a simpler local check

Reply:
It's also used for IPC. I think it's worth having because it's used for at least two protocols. https://github.com/ofiwg/libfabric/blob/e996e60f5d1cfb9253dcb6486ae0a84db14f2500/prov/sm2/src/sm2.h#L327-L336

Unexpected messages in other protocols can't generate completion if delivery complete is requested, yes. I can't think of any other exceptions.

Reply:
I'm ok with merging it as is just to get it in but if we do resume development of sm2 I think we need to rework the completion flag usage. Sometimes the completion is triggered using the delivery complete flag and sometimes it's the generate completion flag and it's set later in the command and then we have this function that is confusing. I think we can make the decision up front with a single flag set and always check that flag.
For example:
```
#define SMR_COMPLETE_LATER

send_atomic()
{
if (!(atomic base && !(op_flags & FI_DELIVERY_COMPLETE))
    xfer_entry->proto_flags |= SMR_COMPLETE_LATER

if (!proto_flags & SMR_COMPLETE_LATER)
    complete_tx() 
}

progress_return()
{
//do return things

if (proto_flags & SMR_COMPLETE_LATER)
    complete_tx()
```

Reply:
Agree. The atomics work was done independently, so we didn't think about using a common flag..